src/components/NotesMindMap/NotesMindMapReducer.js

```diff
@@ -12,116 +12,139 @@ import {
         edges.push({ from: child.parent.id, to: child.id });
       });
-      notes = addGroupTagToNote(notes, state.selectedNote.id);
+      nodes = addGroupTagToNodes(nodes, edges);
     }
-    return { ...state, notes, edges };
+    return {
+      ...state,
+      nodes,
+      edges,
+    };
   };
 
   const handleEditNoteNameAction = () => {
     const noteToEdit = data;
-    return { ...state, showNoteNameEditor: true, selectedNote: noteToEdit };
+    return {
+      ...state,
+      showNoteNameEditor: true,
+      selectedNote: noteToEdit,
+    };
   };
 
   const handleNoteNameUpdateRequestSuccessAction = () => {
     const updatedNote = data;
-    let notes = [...state.notes];
-    notes = notes.filter(note => note.id !== updatedNote.id);
-    notes.push({
-      id: updatedNote.id,
-      label: updatedNote.name,
-      name: updatedNote.name,
-      isNote: updatedNote.isNote,
+    let nodes = [...state.nodes];
+    nodes = nodes.map(node => {
+      return node.id === data.id ? { ...node, name: updatedNote.name } : node;
     });
-    return { ...state, notes };
+    return {
+      ...state,
+      nodes,
+    };
   };
 
   const handleCreateNoteSuccessAction = () => {
     const newNote = data;
-    let notes = [...state.notes];
+    let nodes = [...state.nodes];
     let edges = [...state.edges];
-    notes.push({
-      id: newNote.id,
-      name: newNote.name,
-      label: newNote.name,
-      isNote: newNote.isNote,
-      group: 'children',
-    });
     edges.push({ from: newNote.parent.id, to: newNote.id });
-    notes = addGroupTagToNote(notes, state.selectedNote.id);
-    return { ...state, notes, edges };
+    nodes = addNodeToGraph(nodes, newNote);
+    nodes = nodes.map(node => {
+      return node.id === newNote.parent.id && node.isNote
+        ? { ...node, group: 'parent' }
+        : node;
+    });
+    return {
+      ...state,
+      nodes,
+      edges,
+    };
   };
 
   const handleDeleteNoteRequestSuccessAction = () => {
-    const noteToDelete = data;
-    let notes = [...state.notes];
-    let edges = [...state.edges];
-    let graph = removeNoteFromGraph(notes, edges, noteToDelete);
-    notes = graph.notes;
-    edges = graph.edges;
-    return { ...state, notes, edges };
+    const nodeToDelete = data;
+    const { nodes, edges } = removeNodeFromGraph(
+      state.nodes,
+      state.edges,
+      nodeToDelete,
+    );
+    return {
+      ...state,
+      nodes,
+      edges,
+    };
   };
 
   const handleChangeParentNoteRequestSuccess = () => {
     const noteId = data.noteId;
     const newParentId = data.newParentId;
     let edges = [...state.edges];
-    let notes = [...state.notes];
-    const oldParent = edges.find(edge => edge.to === noteId);
+    let nodes = [...state.nodes];
+    const oldParentId = edges.find(edge => edge.to === noteId).from;
     edges = edges.filter(edge => edge.to !== noteId);
+    const oldParentHasChildren = edges.find(edge => edge.from === oldParentId);
```

Comment:
Why not use a function from the graph helper? https://github.com/DarthVanger/NeuralNotes/pull/80/files#diff-8481f6c43a5e2d6f4ea40d7417e1d388R1

You are duplicating the logic by repeating it here.

Reply:
Done